### ISSUE TITLE
Logging metrics per epoch

### ISSUE THREAD

**Issue by willprice** [mlflow/mlflow#32]
### System information
N/A

### Describe the problem
As it stands, `mlflow` doesn't seem to be capable of logging metrics per epoch during training (e.g. validation metrics during training of big DL models). Are there plans to add support for this and graphing of those metrics (to produce learning curves)?

Many thanks for you excellent project!

EDIT: Looking at https://mlflow.org/docs/latest/tracking.html, the document states:

> Key-value metrics where the value is numeric. Each metric can be updated throughout the course of the run (for example, to track how your model’s loss function is converging), and MLflow will record and let you visualize the metric’s full history.

Looking through the Python API I don't see how this is possible as there is no method to update the current epoch.

**Comment by andrewmchen** [comment]
Thanks for the issue and for the feedback on the APIs. Currently it should be possible to create those learning curves with multiple calls to the `mlflow.log_metric` API and the (get-metrics-history)[https://mlflow.org/docs/latest/rest_api.html#get-metrics-history] API. Moreover, the graphs produced in the UI should graph all values recorded with the `mlflow.log_metric` API.

That being said, currently it is impossible to control the x-value since each call to `log_metric` records only the timestamp at which it was recorded, the metric key, and the metric value. That means it is also impossible to update metric values already recorded.

If being able to control the x-axis or to update values already recorded is important, please let us know!

**Comment by mateiz** [comment]
Just to add onto what Andrew said, you can also see curves of the metrics in the MLflow UI by clicking the metric name on either a Run's individual page or the Compare page between multiple runs. Hopefully that helps. It does look like there's no get-metrics-history in Python though so we need to expose that.

**Comment by willprice** [comment]
Thanks @andrewmchen and @mateiz for the prompt responses, much appreciated! I wasn't aware that the metrics were time-stamped and I'd be able to click on them to see history, that's cool! It'd be nice to log the x-value as well to see how validation metrics change over epochs/steps similar to how comet.ml does it: https://www.comet.ml/docs/python-sdk/Experiment/#experimentlog_metric or https://www.comet.ml/docs/python-sdk/Experiment/#experimentlog_epoch_end

**Comment by mateiz** [comment]
@cyril-p that makes sense! We need to figure out how the user interface for that would look but it's a good idea.

**Comment by 5ke** [comment]
Yes please, to both x-axis control and plotting several metric together! :-)

How about making an interactive graph with plot-activating-legend under the heading metrics (this is a standard feature of plotly, see for example the second graph [here](https://plot.ly/python/create-online-dashboard/#choose-plots))? This way the metrics plot doesn't get burried so deeply and the user is free to select whatever is of relevance. It would require x-axis control, to make sure things get aligned. Y-axis scaling could be automatic (would work in many cases, but wil lead to terrible plots when two metrics of different order of magnitude are plotted together) or by adding an extra axis when required (even more awesome).

**Comment by sotte** [comment]
I just want to chime in. Being able to explicitly log per epoch is a crucial feature. I actually routinely log parameters per epoch (changing learning rate and momentum, anybody?).

**Comment by eyadsibai** [comment]
I am curious to know. the reason you can plot a metric if you cannot have multiple values?

**Comment by mateiz** [comment]
We're actually hoping to add this in the API soon. @smurching is working on a design.

**Comment by zblz** [comment]
@smurching, @mateiz -- as part of the design it might be worth considering also whether artifacts could be attached to a given epoch. It would make sense to tie together, e.g., a serialized model, example generated images, and the metrics for an epoch so that they could be explored and retrieved in a coherent way.

**Comment by jeremyjordan** [comment]
Glad to see this issue is being worked on, I'm anxiously awaiting the ability to control the x-axis. Something as simple as `mlflow.log_metric(name, value, step)` would be perfect.

**Comment by lp-noel** [comment]
Just to add to the discussion, it would be great to be able to plot more than one metric, e.g. training and validation loss. It is a basis for ML models training analysis.

**Comment by smurching** [comment]
@Zangr and @dbczumar have added this functionality to MLflow 1.0 APIs & the UI as described in https://github.com/mlflow/mlflow/blob/master/CHANGELOG.rst#major-features-improvements-and-breaking-changes. See that e.g. [`log_metric`](https://mlflow.org/docs/latest/python_api/mlflow.html#mlflow.log_metric) now accepts an optional `step` parameter for specifying an epoch

will close this for now but please reopen if anything comes up!
